### ISSUE TITLE
Minimal example without three / ikHelper

### ISSUE THREAD

**Issue by felixfaire** [gkjohnson/closed-chain-ik-js#51]
Hi. It would be really helpful to have a minimalistic working example in the readme or the repo. I have tried to create a minimal example of the solver based on the "goals" sample with a basic procedurally moving goal and i cannot get the solver to converge even in the simplest of cases. Copying the "goals" samples line for line without the mouse interaction also seems to lead to "DIVERGED" or "STALLED" in every case. Is there something i am missing?

Here is the code I am trying to get working. Any help appreciated, thanks

```javascript
const solverOptions = {
	maxIterations: 3,
	divergeThreshold: 0.005,
	stallThreshold: 1e-3,
	translationErrorClamp: 0.25,
	rotationErrorClamp: 0.25,
	translationConvergeThreshold: 1e-3,
	rotationConvergeThreshold: 1e-3,
	restPoseFactor: 0.001,
};

let ikRoot = null;
let currRoot = null;

for ( let i = 0; i < 6; i ++ ) {

  const link = new Link();
  const joint = new Joint();
  joint.setPosition( 0.0, 1.0, 0 );
  joint.setDoF( i % 2 ? DOF.EX : DOF.EZ );
  joint.setDoFValues( Math.PI / 4 );
  joint.setRestPoseValues( Math.PI / 4 );
  joint.restPoseSet = true;
  joint.setMinLimits( - 0.9 * Math.PI );
  joint.setMaxLimits( 0.9 * Math.PI );

  link.addChild( joint );

  if ( currRoot ) {
    currRoot.addChild( link );
  }

  if ( ikRoot === null ) {
    ikRoot = link;
  }

  currRoot = joint;
}

const finalLink = new Link();
finalLink.setPosition( 0, 0.5, 0 );
currRoot.addChild( finalLink );

// TODO: rotation seems not to work here
let goal = new Goal();
goal.makeClosure( finalLink );
goal.setGoalDoF(DOF.X, DOF.Y, DOF.Z);

// create solver
const solver = new Solver( [ ikRoot, goal ] );
Object.assign( solver, solverOptions );



..animationLoop() 
{
  goal.setPosition(Math.cos(elapsedTime) * 3.0, 3.0, 1.0);

  let statuses = solver.solve();
  
  let statusString = "";
  statuses.forEach((s, i) => {
    statusString+= SOLVE_STATUS_NAMES[statuses[i]] + ", ";
  })
  console.log(statusString);
}
```

**Comment by gkjohnson** [comment]
Hello! In your example the goal is being moved around pretty far away from the starting point of the ik chain -- not sure if that was intended:

```js
const worldPos = new Float32Array( 3 );
finalLink.getWorldPosition( worldPos );
console.log( worldPos );
// -2.3517, 0.2607, 2.4508

goal.setPosition( Math.cos( 0 ) * 3.0, 3.0, 1.0 );
console.log( goal.position );
// 3.0, 3.0, 1.0
```

There are a lot of factors involved in whether or not a goal can be solved for and it's very difficult to debug these things without some kind of visual representation so I can't tell you exactly why it can't reach that point. It could just not be reachable, the joint limits could be hit, the solvers translationErrorClamp setting could be too large, etc. There's also an issue right now where using the rest pose can cause the solver to "stall" even though it's reached the goal (related to floating point precision issues I believe).

At the very least I would start the goal off at the `finalLink` position, which gives a converged results:

```js
finalLink.getWorldPosition( goal.position );
goal.setMatrixNeedsUpdate();
```

Regarding minimal examples I have one written out in the README [here](https://github.com/gkjohnson/closed-chain-ik-js#simple-2-dof-system) that just sets up an ik system and solver. If you have any suggestions on how to improve it I'd be happy to hear them!

Also can I ask what you're using this for? I'm always interested to hear how people are using my projects.

**Comment by felixfaire** [comment]
Hey thanks again for this.

I had tried several options before creating this issue, trying to use more relaxed Dofs and adding tiny values to the start goal state (starting at the end of the chain) but to no avail. I will look into the translationErrorClamp settings (wasn't sure what they corresponded to).

I am working on a generative graphic design system using various kinetic armatures so the constraints can be extremely relaxed (as the output is purely graphical). I'm using glMatrix and the regl library to create generative visuals and animations from armature motions but had some difficulty with choreographing nontrival armatures so came across your library and appreciated the apparent simplicity and renderer agnosticism.

Regarding the readme example, that was my first port of call but the variable names are inconsistent and when tidying up a bit  (assuming some variable names) i still couldnt get a converged result with a simple chain and a target that I was sure would be in reach (eg 3 links of length 0.5 and a goal point inside a unit sphere).

**Comment by gkjohnson** [comment]
> I had tried several options before creating this issue, trying to use more relaxed Dofs and adding tiny values to the start goal state (starting at the end of the chain) but to no avail.

When placing the goal exactly at the end of the chain I get a `CONVERGED` result. When moving the goal around a small amount relative to the end of the final link I get `CONVERGED` and `STALLED`. Keep in mind that `STALLED` does not mean that the joint chain has not moved towards the goal -- you can see in the example pages that the robot can look like it's satisfied the goal constraints entirely but the result is still returned as `STALLED`. Again this is related to the rest pose. What can happen is the system reaches the goal and then the rest pose is applied which moves the joints slightly away from the converged solution. This _shouldn't_ happen but I suspect that issues related to floating number precision are causing this and I haven't gotten around to looking deeper into it.

Bottom line is that `STALLED` isn't necessarily bad (but it could mean the goals couldn't be reached because joint limits were hit or the goal was out of reach). If you set `restPoseFactor` to 0 you should see the above-mentioned stalling issue go away if it could reach the goal.

> I will look into the translationErrorClamp settings (wasn't sure what they corresponded to).

It's listed in the docs but maybe it's not so clear. When performing a solve the final goal is not jumped to immediately. Instead small steps are made towards to goal because the solver approach is best used for small linear deltas. The `translationErrorClamp` and `rotationErrorClamp` dictate how large that delta can be. For example if the end effector is 1 meter away from the goal and `translationErrorClamp` is 0.1 then the solver will try to move the end effector 0.1 meters closer to the goal rather than 1 meter. Small values will yield more stable solutions but as you can see it means you have to solve over more steps.

> I am working on a generative graphic design system using various kinetic armatures

This sounds like a really interesting project -- let me know once you've made progress! I'll be interested in checking it out.

> Regarding the readme example, that was my first port of call but the variable names are inconsistent

I see `link` was used in place of `link3`. Thanks for the catch. I fixed in in https://github.com/gkjohnson/closed-chain-ik-js/commit/2f8f749bff24467fa7b90e95d1086f38e21bcae6 -- let me know if there are any other inconsistencies I missed!

> I would really love to get this working as it could really elevate the motion of the project so hopefully with some tweaks to settings / goals I can get some solutions. My basic case is a 3 joint arm (akin to a 6axis UR robot) with (i suppose) 2 axis euler Dof at each joint. Any advice would be welcome.

I recommend using the existing three.js preview examples to get a feel for creating an IK system, how to move goals around, and how the solver "responds" to different movements. Large jumps in goal position (from the left side of the arm to the directly opposite right side) are not recommended and possibly cause the solver to diverge or take a path that will butt it up against joint limits. And lastly keep in mind that the reachability of a system with multiple rotation degrees of freedom is not always so intuitive especially when dealing with 6DoF goal constraints so take care when positioning the ik target. I've been meaning to look into visualizing the reachability of an IK system but haven't had the time or the need. Maybe some day!

**Comment by felixfaire** [comment]
Thanks so much for this useful info. I have regularly converging solutions now but still the joints dont seem to move (static, position, quaternion and matrix each frame). To be clear: do you have to manually update the frame matrices each update? I would expect the solve() iterations to do this but looking into the IKRootHelper used in the three examples it seems there is some manual updating required. I have tried calling:

`ikRoot.updateMatrixWorld(true);` &
`finalJoint.updateMatrixWorld(true);`

each frame, but whilst the goal is clearly moving and the solver is reporting converged none of the joint matrices or quaternions appear to change on each frame. Is there something i'm still missing about getting the joint transforms out of the solver each frame (without using three js helpers)?

**Comment by felixfaire** [comment]
For example: in the readme example, after the solve() would it be reasonable to console.log(joint3.position)? Or are the references not updated each solve... (this is similar to how i've been trying to query the joint transforms each solve() / frame).

**Comment by felixfaire** [comment]
Ok with a bit more poking I can see that the worldMatrix of each joint / link updates with the solve but each .matrix .position and .quaternion seem to always be static. Is there a correct way to use the local transform information in the heirarchy?

**Comment by gkjohnson** [comment]
I assume you're trying to get the new, solved local transform relative to the parent of the links / joints in order to update another visual representation. The solver does not adjust the local position or rotation of the joints / links it instead updates the degree of freedom state. Joints have an additional [DoF matrix](https://github.com/gkjohnson/closed-chain-ik-js/blob/main/src/core/Joint.js#L63) which represents the DoF offsets set on the joint and is used to compute the matrix world like so:

```js
this.matrixWorld = parent.matrixWorld * this.matrixDoF * this.matrix
```

If you need the local transformations after the DoF offsets have been applied you'll have to calculate that yourself from `matrixDoF` and `matrix` or using `matrixWorld`.

>  do you have to manually update the frame matrices each update?

After a solve the matrix worlds should be up to date. `updateMatrixWorld` is called defensively in the helpers to account for the situation where a user has made changes without running solve that we want to be visually reflected.

> Is there something i'm still missing about getting the joint transforms out of the solver each frame (without using three js helpers)

I think I answered this above but maybe it's worth noting that the IKRootsHelper classes use matrix world to position all the links:

https://github.com/gkjohnson/closed-chain-ik-js/blob/main/src/three/IKLinkHelper.js#L82

And the same could be done for Joints but due to how they're being visualized there's a custom calculation done without the DoF offsets.

Let me know if that wasn't what you were looking for.

**Comment by felixfaire** [comment]
matrixDof was what i was missing! Its working great now, thanks so much for your help. Coming from various other 3D frameworks / scenegraphs it was slightly counter-intuitive that the .matrix / .quaternion didnt include the relative rotations by default but i can see the logic now.

**Comment by gkjohnson** [comment]
Hey @felixfaire I was just going through a few of the issues to clean up -- did my README update in https://github.com/gkjohnson/closed-chain-ik-js/commit/2f8f749bff24467fa7b90e95d1086f38e21bcae6 solve your issue? Can this be closed? If not what should be added to the README to better demonstrate how to use the library.

Thanks again!
